src/utils/db/car_stream.rs

```diff
@@ -319,96 +364,64 @@ impl<W: AsyncWrite> Sink<CarBlock> for CarWriter<W> {
 }
 
 async fn read_v1_header<ReaderT: AsyncRead + Unpin>(
-    framed_reader: &mut FramedRead<ReaderT, UviBytes>,
-) -> Option<CarV1Header> {
-    let frame = framed_reader.next().await?.ok()?;
-    let header = from_slice_with_fallback::<CarV1Header>(&frame).ok()?;
+    reader: &mut ReaderT,
+) -> std::io::Result<CarV1Header> {
+    let Some(frame) = read_frame(reader).await? else {
+        return Err(std::io::Error::new(
+            std::io::ErrorKind::UnexpectedEof,
+            "failed to decode v1 header frame",
+        ));
+    };
+    let header = from_slice_with_fallback::<CarV1Header>(&frame).map_err(std::io::Error::other)?;
     if header.version != 1 {
-        return None;
+        return Err(std::io::Error::other(format!(
+            "unexpected header version {}, 1 expected",
+            header.version
+        )));
     }
-    Some(header)
+    Ok(header)
+}
+
+async fn read_frame<ReaderT: AsyncRead + Unpin>(
+    reader: &mut ReaderT,
+) -> std::io::Result<Option<Vec<u8>>> {
+    let len: usize = match reader.read_varint_async().await {
+        Ok(len) if len > MAX_FRAME_LEN => {
+            return Err(std::io::Error::new(
+                std::io::ErrorKind::InvalidData,
+                format!("frame too large: {len} > {MAX_FRAME_LEN}"),
+            ));
+        }
+        Ok(len) => len,
+        Err(e) if e.kind() == std::io::ErrorKind::UnexpectedEof => return Ok(None),
+        Err(e) => return Err(e),
+    };
+    let mut bytes = vec![0; len];
+    let n = reader.read_exact(&mut bytes[..]).await?;
+    if n == len {
```

Comment:
nit: This check is not needed, `read_exact` already make sure that we are reading exact `len`